pikelet/src/surface/projections/core.rs

```diff
@@ -432,6 +432,54 @@ pub fn synth_term<S: AsRef<str>>(
                 core_type_value,
             )
         }
+        Term::Let(_, items, term) => {
+            use std::collections::btree_map::Entry;
+            let mut forward_declarations: BTreeMap<&str, Arc<core::Value>> = BTreeMap::new();
+
+            for item in items {
+                match item {
+                    Item::Declaration((range, name, r#type)) => {
+                        if let Entry::Vacant(entry) = forward_declarations.entry(name.as_ref()) {
+                            let (core_type, _core_term) = check_type(state, r#type);
+                            let core_type_value = state.eval_term(&core_type);
+                            entry.insert(core_type_value);
+                        } else {
+                            // There already exists a declaration with this name.
+                            state.report(Message::AmbiguousTerm {
+                                range: range.clone(),
+                                term: AmbiguousTerm::Let,
+                            });
+                            return (core::Term::Error, Arc::new(core::Value::Error));
+                        }
+                    }
+                    Item::Definition((range, name, term)) => {
+                        match forward_declarations.entry(name.as_ref()) {
+                            Entry::Vacant(_entry) => {
+                                // Unannotated, need to try and infer the type?
+                                state.report(Message::AmbiguousTerm {
+                                    range: range.clone(),
+                                    term: AmbiguousTerm::Let,
+                                });
+                                return (core::Term::Error, Arc::new(core::Value::Error));
+                            }
+                            Entry::Occupied(entry) => {
+                                let term = Arc::new(check_term(state, term, entry.get()));
+                                let value = state.eval_term(&term);
+                                state.push_local(
+                                    name.as_ref().to_string(),
+                                    value,
+                                    Arc::clone(entry.get()),
+                                );
+                            }
+                        }
+                    }
+                }
+            }
+
+            // FIXME Not sure if this is really the right thing to do,
+            // presumably we need to pop_many_locals somewhere!
+            synth_term(state, term)
```

Comment:
We'll need to generate let bindings in the core language here. Might be worth discussing how we want to represent them? And then we'll need to write some validation code to check if the elaborated let expressions are well-formed, in `core::typing`.